src/profile-logic/import/art-trace.js

```diff
@@ -0,0 +1,932 @@
+    const low = this.getU32();
+    const high = this.getU32();
+    return high * Math.pow(2, 32) + low;
+  }
+
+  getBytesUntil(endPos) {
+    if (endPos < this._pos) {
+      throw new Error(
+        `getBytesUntil() called with a target position in the past (curPos: ${this._pos}, endPos: ${endPos})`
+      );
+    }
+    const buffer = this._u8View.subarray(this._pos, endPos);
+    this._pos = endPos;
+    return buffer;
+  }
+
+  getBytes(byteCount) {
+    return this.getBytesUntil(this._pos + byteCount);
+  }
+
+  getString(byteLength) {
+    const stringBytes = this.getBytes(byteLength);
+    return this._decoder.decode(stringBytes);
+  }
+
+  // Returns the line without the line break at the end, but advances the
+  // position to after the line break.
+  getLine() {
+    let nextLineBreak = this._u8View.indexOf(lineBreakByte, this._pos);
+    if (nextLineBreak === -1) {
+      nextLineBreak = this._u8View.length;
+    }
+    const line = this._u8View.subarray(this._pos, nextLineBreak);
+    this._pos = nextLineBreak + 1;
+    return this._decoder.decode(line);
+  }
+}
+
+type ArtTraceThread = {|
+  tid: number,
+  threadName: string,
+|};
+
+type ArtTrace = {|
+  summaryDetails: {
+    clock: string,
+    pid?: string,
+  },
+  startTimeInUsecSinceBoot: number,
+  threads: ArtTraceThread[],
+  methods: {|
+    methodId: number,
+    className: string,
+    methodName: string,
+    signature: string,
+  |}[],
+  methodActions: {|
+    tid: number,
+    methodId: number,
+    globalTime: number,
+    threadTime: number,
+    action: 'enter' | 'exit' | 'exit-unroll',
+  |}[],
+|};
+
+function detectArtTraceFormat(
+  traceBuffer: ArrayBuffer
+): 'regular' | 'streaming' | 'unrecognized' {
+  try {
+    const lengthOfExpectedFirstTwoLinesOfSummarySection = '*version\nX\n'
+      .length;
+    const firstTwoLinesBuffer = traceBuffer.slice(
+      0,
+      lengthOfExpectedFirstTwoLinesOfSummarySection
+    );
+    const decoder = new TextDecoder();
+    const firstTwoLinesString = decoder.decode(firstTwoLinesBuffer);
+    if (/\*version\n[1-3]\n/.test(firstTwoLinesString)) {
+      return 'regular';
+    }
+  } catch (e) {
+    // Ignore exception and fall through
+  }
+
+  try {
+    const dataView = new DataView(traceBuffer);
+    const magic = dataView.getUint32(0, true);
+    if (magic === TRACE_MAGIC) {
+      return 'streaming';
+    }
+  } catch (e) {
+    // Ignore exception and fall through
+  }
+
+  return 'unrecognized';
+}
+
+function validateMagicHeader(magicHeader) {
+  if (magicHeader !== TRACE_MAGIC) {
+    const expectedString = `0x${TRACE_MAGIC.toString(16)}`;
+    const gotString = `0x${magicHeader.toString(16)}`;
+    throw new Error(
+      `Unexpected magic header, expected ${expectedString} (little endian), got ${gotString}`
+    );
+  }
+}
+function validateVersion(version) {
+  if (version < 1 || version > 3) {
+    throw new Error(
+      `This code only knows how to parse versions 1 to 3, got version ${version}.`
```

Comment:
Should this error reference the android trace format? Is it possible these errors could be operating on a mislabeled profile type? This could be confusing for a user dropping in non-android file and it not working as expecting. I'm not sure how this error gets surfaced in the UI.

Reply:
Good point. I've added a wrapping try {} catch {} which will mention the ART trace format. Here's an example of the final string that's displayed to the user:

> Error: Unserializing the profile failed: Error: Could not parse the profile array buffer as an ART trace: Error: This code only knows how to parse versions 1 to 3, got version 5.